frame/mosaic/src/decay.rs

```diff
@@ -0,0 +1,99 @@
+use frame_support::pallet_prelude::*;
+use num_traits::{CheckedAdd, CheckedDiv, CheckedMul, CheckedSub, One, Saturating, Zero};
+
+pub trait Decayer<Balance, BlockNumber> {
+	fn checked_decay(
+		&self,
+		amount: Balance,
+		last_decay_block: BlockNumber,
+		current_block: BlockNumber,
+	) -> Option<Balance>;
+
+	fn full_recovery_period(&self, amount: Balance) -> Option<BlockNumber>;
+}
+
+/// Recommend type for storing the decay function of a penalty.
+#[derive(Decode, Encode, TypeInfo, Debug, PartialEq, Clone)]
+pub enum BudgetPenaltyDecayer<Balance, BlockNumber> {
+	/// Linear variant of the decay function, which decreases every block.
+	Linear(LinearDecay<Balance, BlockNumber>),
+}
+
+impl<Balance, BlockNumber> BudgetPenaltyDecayer<Balance, BlockNumber> {
+	#[allow(dead_code)]
+	pub fn linear(n: Balance) -> BudgetPenaltyDecayer<Balance, BlockNumber> {
+		BudgetPenaltyDecayer::Linear(LinearDecay { factor: n, _marker: PhantomData })
+	}
+}
+
+impl<Balance, BlockNumber> Decayer<Balance, BlockNumber>
+	for BudgetPenaltyDecayer<Balance, BlockNumber>
+where
+	BlockNumber: CheckedSub + Saturating + Into<Balance> + TryFrom<Balance> + One + CheckedAdd,
+	Balance: CheckedMul + CheckedDiv + Saturating + Zero,
+{
+	fn checked_decay(
+		&self,
+		amount: Balance,
+		last: BlockNumber,
+		current: BlockNumber,
+	) -> Option<Balance> {
+		match self {
+			BudgetPenaltyDecayer::Linear(lin) => lin.checked_decay(amount, last, current),
+		}
+	}
+
+	fn full_recovery_period(&self, amount: Balance) -> Option<BlockNumber> {
+		match self {
+			BudgetPenaltyDecayer::Linear(lin) => lin.full_recovery_period(amount),
+		}
+	}
+}
+
+#[derive(Decode, Encode, TypeInfo, Default, Debug, PartialEq, Clone)]
+pub struct LinearDecay<Balance, BlockNumber> {
+	/// Factor by which we decay every block.
+	factor: Balance,
```

Comment:
hm, this is not relative to block? why?

Reply:
please check time.rs for examples of decays https://github.com/ComposableFi/composable/blob/ff39c92f7eabfa3218adfb252cef89ba7b3593f3/frame/composable-traits/src/time.rs#L21 too

Reply:
CU-1wana2u